File: 【大三】网络与信息安全技术/实验/code/RSA.cpp
#include <iostream>
using namespace std;

long long modexp(long long base, long long exp, long long mod) {
  long long res = 1;
  base %= mod;
  while (exp > 0) {
    if (exp % 2) res = (res * base) % mod;
    base = (base * base) % mod;
    exp /= 2;
  }
  return res;
}

int main() {
  long long p = 61, q = 53;
  long long n = p * q; // 3233
  long long e = 17;
  long long d = 2753;

  long long m = 65;

  long long c = modexp(m, e, n);
  long long decrypted = modexp(c, d, n);

  cout << "n=" << n << endl;
  cout << "Public key: (" << e << "," << n << ")" << endl;
  cout << "Private key: (" << d << "," << n << ")" << endl;
  cout << "Encrypted: " << c << endl;
  cout << "Decrypted: " << decrypted << endl;
}

File: 【大三】网络与信息安全技术/实验/code/RSA.js
// @ts-nocheck

function modexp(base, exp, mod) {
  let result = 1n
  base = BigInt(base) % BigInt(mod)
  exp = BigInt(exp)
  mod = BigInt(mod)

  while (exp > 0n) {
    if (exp % 2n === 1n) result = (result * base) % mod
    base = (base * base) % mod
    exp /= 2n
  }
  return result
}

const p = 61n
const q = 53n
const n = p * q
const e = 17n
const d = 2753n

const m = 65n

const c = modexp(m, e, n)
const decrypted = modexp(c, d, n)

console.log("n =", n.toString())
console.log("Public key:", `(${e}, ${n})`)
console.log("Private key:", `(${d}, ${n})`)
console.log("Encrypted:", c.toString())
console.log("Decrypted:", decrypted.toString())
